node: cpumanager: add options to reject non SMT-aligned workload

swatisehgal opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
We propose a change in [cpumanager](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/375-cpu-manager) to make the behaviour of latency-sensitive applications more predictable when running on SMT-enabled systems. The existing static policy is allowed to allocate single cpus (aka virtual threads on SMT-enabled systems), and this can cause physical core sharing which in turn is a major cause for noisy neighbours, which harms the behaviour of latency-sensitive applications.

We add extra constraints to the aformenentioned policy, while preserving all its properties, to avoid physical core sharing.

#### Which issue(s) this PR fixes:
Fixes # N/A

#### Special notes for your reviewer:
This PR contains E2E tests as well.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.
- KEP PR: https://github.com/kubernetes/enhancements/pull/2626
- Enhancement Issue: https://github.com/kubernetes/enhancements/issues/2625

swatisehgal commented:
/cc @fromanirh

fromanirh commented:
For some more context, on April 20 sig-node meeting we presented this work, but also talked in general about how to extend cpumanager. We thought about demonstrating a prototype, but yhis change is pretty small and self-contained, so we're presenting the actual complete PR alongside the the KEP (https://github.com/kubernetes/enhancements/pull/2626).

klueska commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -477,3 +474,52 @@ func (m *manager) updateContainerCPUSet(containerID string, cpus cpuset.CPUSet)
 	return m.state.GetCPUSetOrDefault(podUID, containerName)
 }
+
+func (m *manager) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
+	klog.InfoS("CPUManager Admit Handler")
+	pod := attrs.Pod
+	return m.policy.Admit(pod)
+}
+
+func smtAlignmentError(requestedCPUs, cpusPerCore int) lifecycle.PodAdmitResult {
+	errorMessage := fmt.Sprintf("Number of CPUs requested should be a multiple of number of CPUs on a core = %d on this system. Requested CPU count = %d", cpusPerCore, requestedCPUs)
+	return lifecycle.PodAdmitResult{
+		Message: errorMessage,
+		Reason:  "SMTAlignmentError",
+		Admit:   false,
+	}
+}
+
+func admitPod() lifecycle.PodAdmitResult {
+	return lifecycle.PodAdmitResult{
+		Admit: true,
+	}
+}
```
I don't think we want a new `Admit` handler here.

We can consider reworking the way `Allocate()` propagates errors to the existing `Admit` handler, but I'm not in favor of adding a new one like this.

klueska commented on pkg/kubelet/cm/cpumanager/policy_smtaware.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+func (p *smtAwarePolicy) Admit(pod *v1.Pod) lifecycle.PodAdmitResult {
+	cpusPerCore := p.topology.CPUsPerCore()
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		if requestedCPUs := p.guaranteedCPUs(pod, &container); requestedCPUs > 0 && (requestedCPUs%cpusPerCore) != 0 {
+			// Since CPU Manager has been enabled with `smtaware` policy, it means a guaranteed pod can only be admitted
+			// if the CPU requested is a multiple of the number of virtual cpus per physical cores.
+			// In case CPU request is not a multiple of the number of virtual cpus per physical cores the Pod will be put
+			// in Failed state, with SMTAlignmentError as reason. Since the allocation happens in terms of physical cores
+			// and the scheduler is responsible for ensuring that the workload goes to a node that has enough CPUs,
+			// the pod would be placed on a node where there are enough physical cores available to be allocated.
+			// Just like the behaviour in case of static policy, takeByTopology will try to first allocate CPUs from the same socket
+			// and only in case the request cannot be sattisfied on a single socket, CPU allocation is done for a workload to occupy all
+			// CPUs on a physical core. Allocation of individual threads would never have to occur.
+			return smtAlignmentError(requestedCPUs, cpusPerCore)
+		}
+	}
```
I think this logic is better served as an error in the call to `Allocate()` (which then triggers an admission error since no allocation can be made).

swatisehgal replied:
Our initial approach was indeed to perform this check in the call to `Allocate` but pivoted to do this at admission stage because it implicitly meant that we would be admitting pods that could result in `SMTAlignment` error eventually. Having said that, we have no issues moving back to performing the check in Allocate but wanted to explain the reason we thought performing in Admit was better.

klueska replied:
It seems the only advantage having an explicit `Admit()` handler gives us, is the ability to more precisely set the “Reason”, which I think could be fixed by updating the way Allocate() propagates its errors back.

I’d also be fine *only* having an `Admit()` handler and not have failures to `Allocate()` trigger an admission error (but then I’d want that to be consistent across all of the components that currently rely on `Allocate()` for this). 

It just seems weird to have an explicit `Admit()` handler for `SMTAlignment`, and then leave all other admission errors to the `Allocate()` call.

swatisehgal replied:
Got it! Thanks for the clarification.

klueska reviewed (commented):
Left a few comments. If you follow my suggestions, I think the surface error of the change could actually be reduced quite a bit.

That said -- how do you plan to expose this policy to the end user?
As a new value to the `--cpu-manager-policy` kubelet flag?
None of those changes seem to be part of this PR.

I'm asking becuase...

This is still technically a `static` cpu-manager-policy. This name is meant to represent the fact that the `cpuset` granted to pods in the guaranteed QoS class are static, i.e. never change. Your new policy still meet that criteria, but just tweaks it slightly to force whole numbered CPUs to be requested.

As such, does it maybe make sense to introduce a new flag for something like this?
At NVIDIA, we also have a small tweak to the static policy on our internal fork, which I never proposed upstream because I didn't know best to fit it into the existing framework.

If we had a new flag (or a way of attaching attributes to the existing flag), I think that would work alot better. It would also mean that we wouldn't need to define a whole new policy, but instead pass an `attributes` struct to the existing `static` policy with knobs for how to customize it.

fromanirh commented:
> Left a few comments. If you follow my suggestions, I think the surface error of the change could actually be reduced quite a bit.

Thanks Kevin for the quick feedback!

> That said -- how do you plan to expose this policy to the end user?
> As a new value to the `--cpu-manager-policy` kubelet flag?
> None of those changes seem to be part of this PR.
> 
> I'm asking becuase...
> 
> This is still technically a `static` cpu-manager-policy. This name is meant to represent the fact that the `cpuset` granted to pods in the guaranteed QoS class are static, i.e. never change. Your new policy still meet that criteria, but just tweaks it slightly to force whole numbered CPUs to be requested.

That's a very good point. Let me take a step back to fully answer.

When we first received feedback about this work, you raised the question about what's the best option to extend cpumanager. And the majority of the feedback received from the community was in the same direction: let's check first what's the direction forward for cpumanager (and, soon emerged, to extend resource managers in general - memory and topology).

So we discussed this in sig-node meeting on April 20, and we dedicated a [good chunk discussion](https://github.com/fromanirh/fromanirh/blob/main/docs/presentations/k8s-cpumanager-smtawareness/smt-sware-cpumanger-sig-node-20210420.pdf) about sorting the extension approach first.

While there it was general consensus that allowing external policies is a nice property, various concerns emerged about (non-comprehensive list): state management, API surface (including the stability promises), how properly delegate core resource management (cpu, memory) to external components, the reliability of the aforementioned components (what if the "cpus" allocator enters a crashloop?).
It seems there is a consensus forming towards [moving the resource management responsabilities into the runtime](https://kccnceu2021.sched.com/event/iE1Y) and towards [NRI](https://github.com/containerd/nri).

So from our (proponents of this change) perspective it makes sense to narrow it down as much as possible to fit in the current framework, with focused minimal changes, which is the motive for the current PR.

Now, answering your question: we are indeed proposing a new policy which totally wants to be a specialized version of the static policy. The best option seemed to be to encapsulate the changes in a new (as much as possible) self-contained policy. But we have some new facts to consider now! (see below).

> As such, does it maybe make sense to introduce a new flag for something like this?
> At NVIDIA, we also have a small tweak to the static policy on our internal fork, which I never proposed upstream because I didn't know best to fit it into the existing framework.
> 
> If we had a new flag (or a way of attaching attributes to the existing flag), I think that would work alot better. It would also mean that we wouldn't need to define a whole new policy, but instead pass an `attributes` struct to the existing `static` policy with knobs for how to customize it.

All considered, this seems a very nice idea. Adding minimal changes and adding attributes to the existing static policy to fine tune its behaviour (which is totally what we want to do here) seems a nice evolution of the existing framework which also plays nicely which the emerging future plans.

Let's talk about the UI though. For our usecase, a new option like
```
--cpu-manager-policy-options=[flag1],[flag2],[...flagn]
```
seems sufficient. For this work it could look  like
```
--cpu-manager-policy-options=smtaware
```
Do you believe this UI would be sufficient to cover the use case(s) you have in NVIDIA?

I'll comment the other points you raised shortly.

(cc @kad and @swatisehgal)

klueska commented:
> ```
> --cpu-manager-policy-options=[flag1],[flag2],[...flagn]
> ```
> seems sufficient. For this work it could look like
> ```
> --cpu-manager-policy-options=smtaware
> ```
> Do you believe this UI would be sufficient to cover the use case(s) you have in NVIDIA?

This would work for us. I fact, your change + ours could work together (which is why I think it makes more sense as a set of attributes than a new policy).

I'm OK with it being a new flag like this, though if we were to go back in time, I would likely add *all* `CPUManager` options as as entries in a `KubeletConfiguration` (and not have any actual flags for them):
```
cpuManager:
    ReconcilePeriod: 
    Policy:
        Name:
        Options:
```

Maybe it's still an option to do something like this instead? I'm not that up-to-speed on what the best-practices around new `kubelet` flags / options are these days. Maybe someone else with more familiarity can comment.

fromanirh commented:
> > ```
> > --cpu-manager-policy-options=[flag1],[flag2],[...flagn]
> > ```
> > 
> > 
> > seems sufficient. For this work it could look like
> > ```
> > --cpu-manager-policy-options=smtaware
> > ```
> > 
> > 
> > Do you believe this UI would be sufficient to cover the use case(s) you have in NVIDIA?
> 
> This would work for us. I fact, your change + ours could work together (which is why I think it makes more sense as a set of attributes than a new policy).
> 
> I'm OK with it being a new flag like this, though if we were to go back in time, I would likely add _all_ `CPUManager` options as as entries in a `KubeletConfiguration` (and not have any actual flags for them):
> 
> ```
> cpuManager:
>     ReconcilePeriod: 
>     Policy:
>         Name:
>         Options:
> ```
> 
> Maybe it's still an option to do something like this instead? I'm not that up-to-speed on what the best-practices around new `kubelet` flags / options are these days. Maybe someone else with more familiarity can comment.

This is a fair point and something can explore. However I believe (and expect) that changing back existing options breaks the backward compatibility, so our options are likely limited here. But let's check indeed.

swatisehgal committed:
smtalign: add cpu-manager-policy-options flag in Kubelet

In this patch we enhance the kubelet configuration to support
cpuManagerPolicyOptions.

In order to introduce SMT-awareness in CPU Manager, we introduce a
new flag in Kubelet to allow the user to specify an additional flag
called `cpumanager-policy-options` to allow the user to modify the
behaviour of static policy to strictly guarantee allocation of whole
core.

Co-authored-by: Francesco Romani <fromani@redhat.com>
Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

ehashman commented:
Putting this in "Waiting on Author" as tests are not passing.

klueska commented on pkg/kubelet/apis/config/fuzzer/fuzzer.go:
```diff
@@ -19,7 +19,7 @@ package fuzzer
 	"time"
 
-	"github.com/google/gofuzz"
+	fuzz "github.com/google/gofuzz"
```
Can we drop this change?

fromanirh replied:
Yep, we'll keep the changes minimal. Looks like these changes are made by overhelpful tooling, so they can surface again, but we will check and strip them down.

klueska commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -762,6 +763,13 @@ func (m *resourceAllocator) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle
 		if m.cpuManager != nil {
 			err = m.cpuManager.Allocate(pod, &container)
 			if err != nil {
+				if _, ok := err.(*topologymanager.SMTAlignmentError); ok {
+					return lifecycle.PodAdmitResult{
+						Message: err.Error(),
+						Reason:  "SMTAlignmentError",
+						Admit:   false,
+					}
+				}
```
I'm fine with this as is, but I'm curious how pervasive the use of "error wrapping" is in the Kubernetes code base: https://blog.golang.org/go1.13-errors#TOC_3.3.

This would be a cleaner solution, but I don't want to introduce something not widely used in K8s yet.

klueska commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -79,6 +83,9 @@ type staticPolicy struct {
 	cpusToReuse map[string]cpuset.CPUSet
+	// flag to enable extra allocation restrictions to avoid
+	// different containers to possibly end up on the same core
+	smtAwarenessEnabled bool
```
Can we create an `Options` struct to wrap these instead of a single variable for each.

klueska commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -29,8 +29,12 @@ import (
+	// PolicyStatic is the name of the static policy
+	PolicyStatic policyName = "static"
+	// SMTAwarePolicyOption is the name of the CPU Manager policy option
+	SMTAwarePolicyOption string = "smtaware"
```
I probably should have brought this up earlier, but I'm not sold on the name `smtaware` for this option.

The static policy is already technically `smtaware`.  In fact, this option actually leverages the fact that it is already `smtaware` to ensure that it allocates all hyper-threads from a single-cpu before considering hypethreads from other cpus.

This option simply add a restriction on top of this to only admit pods that request cores in multiples of the number of hyperthreads per cpu (thus forcing the existing policy to allocate whole-cpus).

I don't have a good suggestion for a different name off the top of my head, but I think we want something more descriptive than just `smtaware`.

fromanirh replied:
Yes, this is totally true and a valid point. `smtaware` is the internal name we picked up initially when we started researching and of course it stuck (naming is hard).
I think we can just replace `smtaware` with something more accurate like `smtforce` or `smtalign` or `smtforcealign`.
Thinking about it, we will need to check if the user-facing value shoud be `camelCase` (`SMTForceAlign`) or not.

klueska commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -109,10 +116,11 @@ func NewStaticPolicy(topology *topology.CPUTopology, numReservedCPUs int, reserv
-		topology:    topology,
-		reserved:    reserved,
-		affinity:    affinity,
-		cpusToReuse: make(map[string]cpuset.CPUSet),
+		topology:            topology,
+		reserved:            reserved,
+		affinity:            affinity,
+		cpusToReuse:         make(map[string]cpuset.CPUSet),
+		smtAwarenessEnabled: isSMTAwarePolicyOptionEnabled(cpuPolicyOptions),
 	}, nil
```
Assuming we introduce an `Options` struct, as suggested above -- I think we should have a Constructor for it that takes in the options in `[]string` form and returns a populated `Options` struct that can be assigned here.

Likewise, this constructor should also return an error indicating whether the incoming list of strings are valid options or not.

Something like:
```
type Options struct {
    smtAwareness bool
    ...
}

func NewOptions([]string{}) (Options, error) {
   ...
}
```

klueska commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -504,3 +527,12 @@ func (p *staticPolicy) generateCPUTopologyHints(availableCPUs cpuset.CPUSet, reu
 	return hints
 }
+
+func isSMTAwarePolicyOptionEnabled(policyOptions []string) bool {
+	for _, option := range policyOptions {
+		if option == SMTAwarePolicyOption {
+			return true
+		}
+	}
+	return false
+}
```
This would likely either become code that sits directly inside the `Options` constructor, or a private function hanging off the `Options` type.

fromanirh replied:
ack, I think we should start putting this logic in the constructor.

klueska commented on pkg/kubelet/cm/topologymanager/scope.go:
```diff
@@ -21,7 +21,7 @@ import (
 	"sync"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Can we drop this?

fromanirh replied:
yep, same as https://github.com/kubernetes/kubernetes/pull/101432#discussion_r624933346

klueska replied:
Looks like this (and similar below) out-of-scope changes are still part of the PR?

fromanirh replied:
yes, because the overzealous `gopls` on behalf of `vim-go` keeps adding them. Will have a remove pass. Sorry for the churn.

klueska commented on pkg/kubelet/cm/topologymanager/scope.go:
```diff
@@ -162,3 +165,20 @@ func unexpectedAdmissionError(err error) lifecycle.PodAdmitResult {
 func admitPod() lifecycle.PodAdmitResult {
 	return lifecycle.PodAdmitResult{Admit: true}
 }
+
+func smtAlignmentError(err error) lifecycle.PodAdmitResult {
+	return lifecycle.PodAdmitResult{
+		Message: err.Error(),
+		Reason:  "SMTAlignmentError",
+		Admit:   false,
+	}
+}
+
+type SMTAlignmentError struct {
+	RequestedCPUs int
+	CpusPerCore   int
+}
+
+func (e SMTAlignmentError) Error() string {
+	return fmt.Sprintf("Number of CPUs requested should be a multiple of number of CPUs on a core = %d on this system. Requested CPU count = %d", e.CpusPerCore, e.RequestedCPUs)
+}
```
It feels strange to me that this error type is defined here and not in the `CPUManager`.

fromanirh replied:
It was done because it was the simplest way to avoid circular imports, but yes, it should be inside the `CPUManager` somehow. We'll think about a better approach.

klueska commented on pkg/kubelet/cm/topologymanager/scope.go:
```diff
@@ -125,6 +125,9 @@ func (s *scope) admitPolicyNone(pod *v1.Pod) lifecycle.PodAdmitResult {
 	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
 		err := s.allocateAlignedResources(pod, &container)
 		if err != nil {
+			if _, ok := err.(*SMTAlignmentError); ok {
+				return smtAlignmentError(err)
+			}
```
It's unfortunate that this needs to be type checked like this in so many places (I count at least 4). Is there some better way to avoid this?

fromanirh replied:
This is potentially related to https://github.com/kubernetes/kubernetes/pull/101432#discussion_r623354087 .
The reason why this block is replicated is because there indeed are 4 places on which we do the admission control:
1. container manager (topology manager disabled)
2. topology manager, when enabled, delegated by container manager; if the policy is none, then topology manager itself does the admission controls otherwise it further delegates to the 
3. scope_container in topology manager or
4. scope_pod in topology manager

Additionally, `SMTAlignmentError` bleongs in cpumanager, but the specific resource managers need to import the `topologymanager` package because of the `HintProvider` interface.

A way to improve both sides could be to run the actual admission check in the resource manager, but to build the actual `lifecycle.PodAdmitResult` always in a single place, and the most suitable seems to be the container manager.
IOW, the container manager will keep delegating the actual checks - which the specific resource managers know how to do - but will centralize the admission control proper.

This way we deduplicate all the `lifecycle.PodAdmitResult` creation bits and we have a cleaner import hierarchy - container manager needs to import both `topologymanager` and `cpumanager` already!
However unfortunately this change will require some refactoring to fix the managers interface. I prototyped this change here: https://gist.github.com/fromanirh/33d2094a8fdf23fdcaf9e38f7003bb28

Could be this a good direction forward?

klueska commented on pkg/kubelet/cm/topologymanager/scope_container.go:
```diff
@@ -17,7 +17,7 @@ limitations under the License.
 
 import (
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Can we drop this?

fromanirh replied:
yep, same as https://github.com/kubernetes/kubernetes/pull/101432#discussion_r624933346

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

klueska commented:
Does it make sense to split this into two PRs? One for the API change to add `--cpu-manager-policy-options`. And one to add the actual option itself. That way (once the API change is approved), I will have full approval rights for the rest of the changes and we can iterate more quickly, knowing it won't be blocked by the API change.

swatisehgal commented:
> Does it make sense to split this into two PRs? One for the API change to add `--cpu-manager-policy-options`. And one to add the actual option itself. That way (once the API change is approved), I will have full approval rights for the rest of the changes and we can iterate more quickly, knowing it won't be blocked by the API change.

Sounds like a good idea!

ffromani committed:
smtalign: propagate policy options to cpumanager

The CPUManagerPolicyOptions received from the kubelet config/command line args
is propogated to the Container Manager.

We defer the consumption of the options to a later patch(set).

Co-authored-by: Swati Sehgal <swsehgal@redhat.com>
Signed-off-by: Francesco Romani <fromani@redhat.com>

klueska commented:
I think if you just take your first 4 commits plus this one line from the 5th, that would make a good isolated commit for the API change:
```
diff --git a/pkg/kubelet/cm/cpumanager/policy_static.go b/pkg/kubelet/cm/cpumanager/policy_static.go
index c7b3172ea4f..4ecaa7d6810 100644
--- a/pkg/kubelet/cm/cpumanager/policy_static.go
+++ b/pkg/kubelet/cm/cpumanager/policy_static.go
@@ -87,7 +87,7 @@ var _ Policy = &staticPolicy{}
 // NewStaticPolicy returns a CPU manager policy that does not change CPU
 // assignments for exclusively pinned guaranteed containers after the main
 // container process starts.
-func NewStaticPolicy(topology *topology.CPUTopology, numReservedCPUs int, reservedCPUs cpuset.CPUSet, affinity topologymanager.Store) (Policy, error) {
+func NewStaticPolicy(topology *topology.CPUTopology, numReservedCPUs int, reservedCPUs cpuset.CPUSet, affinity topologymanager.Store, cpuPolicyOptions []string) (Policy, error) {
    allCPUs := topology.CPUDetails.CPUs()
    var reserved cpuset.CPUSet
    if reservedCPUs.Size() > 0 {
```

fromanirh commented:
> I think if you just take your first 4 commits plus this one line from the 5th, that would make a good isolated commit for the API change:
> 
> ```
> diff --git a/pkg/kubelet/cm/cpumanager/policy_static.go b/pkg/kubelet/cm/cpumanager/policy_static.go
> index c7b3172ea4f..4ecaa7d6810 100644
> --- a/pkg/kubelet/cm/cpumanager/policy_static.go
> +++ b/pkg/kubelet/cm/cpumanager/policy_static.go
> @@ -87,7 +87,7 @@ var _ Policy = &staticPolicy{}
>  // NewStaticPolicy returns a CPU manager policy that does not change CPU
>  // assignments for exclusively pinned guaranteed containers after the main
>  // container process starts.
> -func NewStaticPolicy(topology *topology.CPUTopology, numReservedCPUs int, reservedCPUs cpuset.CPUSet, affinity topologymanager.Store) (Policy, error) {
> +func NewStaticPolicy(topology *topology.CPUTopology, numReservedCPUs int, reservedCPUs cpuset.CPUSet, affinity topologymanager.Store, cpuPolicyOptions []string) (Policy, error) {
>     allCPUs := topology.CPUDetails.CPUs()
>     var reserved cpuset.CPUSet
>     if reservedCPUs.Size() > 0 {
> ```

Yep, thanks! we're working in this direction! :)

fromanirh commented:
this is the full PR, we will split these commits as suggested on a separate PR ASAP:
```
97959bd3543 smtalign: propagate policy options to cpumanager
63b5a0cb307 smtalign: resolve API Rule Check Failure
5a43c1f37c5 smtalign: add auto generated code
90cd0951146 smtalign: add cpu-manager-policy-options flag in Kubelet
```

So this PR will contain these commits only:
```
df0d351cd3a  smtalign: e2e: add tests
0bdb12ed744 smtalign: allow SMT-alignment in CPU Manager
79e3b6ecb5b smtalign: propagate policy options to static policy
020786a302d smtalign: cm: centralize admission handler
```

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

fromanirh commented:
/test pull-kubernetes-e2e-kind-ipv6

klueska commented on pkg/kubelet/cm/admission/errors.go:
```diff
@@ -0,0 +1,58 @@
+limitations under the License.
+*/
+
+package admission
```
I like the idea of having a dedicated `admission` package, but I think there is a better way to handle this abstraction that doesn't require pulling in `topologymanager`, `cpumanager`, etc. here. It's easier if I just propose a patch on top of this than try and explaining everything in text. I will send you something later today or tomorrow to see what you think.

klueska replied:
Here it is as a diff to the tip of your PR:
https://github.com/klueska/kubernetes/compare/pr/101432%5E...pr/101432

Much of it is a "revert" of many of the changes you had to make to introduce your original package.

fromanirh replied:
It is way nicer indeed. I thought about extending the `Error` interface somehow but I stopped short because I could only think of what would end up to be a poor man alternative of type assertion, and came with the approach I propsed. But your solution avoid that direction completely. I'll switch to your direction.

klueska replied:
Can you use my `kklues@nvidia.com` email for the attribution please?

fromanirh replied:
Sure thing; will push again soon anyway.

ffromani committed:
smtalign: cm: factor out admission response

Introduce a new `admission` subpackage to factor out the responsability
to create `PodAdmitResult` objects. This enables resource manager
to report specific errors in Allocate() and to bubble up them
in the relevant fields of the `PodAdmitResult`.

To demonstrate the approach we refactor TopologyAffinityError as a
proper error.

Co-authored-by: Kevin Klues <kklues@nvidia.com>
Co-authored-by: Swati Sehgal <swsehgal@redhat.com>
Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
New upload to address the feature gate addition (see https://github.com/kubernetes/enhancements/pull/2626#issuecomment-833611063 for context)

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

klueska commented on pkg/kubelet/cm/cpumanager/options.go:
```diff
@@ -0,0 +1,43 @@
+/*
```
Can we call this file `policy_options.go`

fromanirh replied:
ack

klueska commented on pkg/kubelet/cm/cpumanager/options.go:
```diff
@@ -0,0 +1,43 @@
+	SMTAlignPolicyOption string = "smtalign"
+)
+
+type Options struct {
```
I think this should probably have a more specific name. At least `PolicyOptions` if not actually full-on `StaticPolicyOptions`.

fromanirh replied:
Going for `StaticPolicyOptions`

klueska commented on pkg/kubelet/cm/cpumanager/options.go:
```diff
@@ -0,0 +1,43 @@
+	SMTAlignmentEnabled bool
+}
+
+func NewOptions(policyOptions []string) (Options, error) {
```
This should definitely be very specific to `NewStaticPolicyOptions()` so that it filters down to exactly those options accepted by the `static` policy.

fromanirh replied:
ack

klueska commented on pkg/kubelet/cm/cpumanager/options_test.go:
```diff
@@ -0,0 +1,90 @@
+/*
```
I would just add these to the existing `policy_static_test.go` file.

fromanirh replied:
ack

klueska commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -230,6 +248,21 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 		klog.InfoS("Static policy: Allocate", "pod", klog.KObj(pod), "containerName", container.Name)
 		// container belongs in an exclusively allocated pool
 
+		if p.options.SMTAlignmentEnabled && numCPUs%p.topology.CPUsPerCore() != 0 {
```
Can we add some parentheses here to make this slightly more clear, i.e.
```
		if p.options.SMTAlignmentEnabled && ((numCPUs%p.topology.CPUsPerCore()) != 0) {

```

fromanirh replied:
ack

fromanirh commented:
rebased

fromanirh commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

klueska commented on cmd/kubelet/app/options/options.go:
```diff
@@ -482,6 +482,7 @@ func AddKubeletConfigFlags(mainfs *pflag.FlagSet, c *kubeletconfig.KubeletConfig
 	fs.StringVar(&c.CgroupDriver, "cgroup-driver", c.CgroupDriver, "Driver that the kubelet uses to manipulate cgroups on the host.  Possible values: 'cgroupfs', 'systemd'")
 	fs.StringVar(&c.CgroupRoot, "cgroup-root", c.CgroupRoot, "Optional root cgroup to use for pods. This is handled by the container runtime on a best effort basis. Default: '', which means use the container runtime default.")
 	fs.StringVar(&c.CPUManagerPolicy, "cpu-manager-policy", c.CPUManagerPolicy, "CPU Manager policy to use. Possible values: 'none', 'static'.")
```
I guess this line should be updated with the new policy name once we decide on it -- and then push that new policy name all the way through the other files.

swatisehgal replied:
Right, we will update.

klueska commented on pkg/kubelet/cm/cpumanager/policy_options.go:
```diff
@@ -0,0 +1,43 @@
+
+const (
+	// RejectNonSMTAlignedOption is the name of the CPU Manager policy option
+	RejectNonSMTAlignedOption string = "reject-non-smt-aligned"
```
I've been thinking about this for a while, and I think I like the name `full-cpus-only` for this option. It describes the intent appropriately, and it leaves it open for composability with other options that may build on this idea beyond just using it to decide to reject a pod with containers requesting an inappropriate number of cores to satisfy this constraint.

swatisehgal replied:
I like the idea but was thinking that in Kubernetes context, the term CPU can translate to a virtual core or a physical core depending on whether hyperthreading is enabled on a node or not. Because of this `full-cpus-only` could be misinterpreted. How about `full-physical-core-only` or `full-core-only`?

klueska replied:
Hmm. That's true. Kubernetes does tend to use `cpu` as a general term that could mean either `cpu` or `core` in the traditional sense (i.e. where the hierarchy goes `sockets -> cpus -> cores`). Linux also doesn't play nicely with this naming hierarchy (at least not when it comes to looking in `/proc/cpuinfo`). 

With that in mind, the names `full-physical-core-only` or `full-core-only` also seem inappropriate because they could also lead to confusion (for example, the lowest level entity in my mental model is always a core, so talking about a "full" core is meaningless to me).

swatisehgal replied:
Well, the terminology is a bit confusing which makes naming challenging but the reason I suggested `full-physical-core-only` or `full-core-only` because my understand is that the hierarchy is `sockets -> cores -> cpus`. In case of a system with the following configuration:
```
Thread(s) per core:   2 (HT enabled)
Core(s) per socket:   12
Socket(s):            4
```
it has 4 CPU sockets, each socket has 12 cores and each core can have two threads (which I was referring to as CPUs as they are the execution units).

klueska replied:
I guess that's where the confusion is though.

Much of the literature out there (and the way I've always referred to it in my 20+ years of working with this) follows the hierarchy `sockets -> cpus -> cores`. Just do a quick google search for "[cpus vs. cores](https://www.google.com/search?q=cores+vs.+cpus)" and see what comes up.

Unfortunately, the problem seems to be that these terms have always been somewhat ambiguous (and never really agreed upon between the OS and the architecture communities), leading to this confusion. In fact, I always forget that Linux actually swaps theses terms from how I think about them, as you pointed out above.

Doing a quick google search, it appears BSD is similar though:
```
FreeBSD/SMP: Multiprocessor System Detected: 24 CPUs
FreeBSD/SMP: 2 package(s) x 6 core(s) x 2 SMT threads
```

It could very well be that I'm the odd one out and just stuck in my old ways. I'm curious what others think though. Has there been a convergence around these terms contrary to my understanding?

swatisehgal replied:
Great point! Sure, let's see what others think on this subject.

fromanirh replied:
My mental model was quite similar to @klueska 's. Problem is: it seems, like @swatisehgal pointed out, that kubernetes is naming thing in yet another slightly different way than OS community historically did and than HW manufacturers are doing.
In out case, I think is better to align as much as possible to k8s' naming, even though the naming choice k8s did are debatable in the first place (at least IMHO).

That said, going back to the `sockets -> cpus -> cores` model, while I share a similar background, in recent times we are having a rapidly evolving landscape on which new terms (or different meaning for the same terms) are more and more common, for example `packages` or `dies` or `chips` or even fancier concepts like `cpu complexes` (which is really what we could have called `cores` + `cache` + `i/o link` ?). Even naming the entity we refer as "virtual core" is challenging, should it a "virtual core", a "execution unit/port", a "hyperthread", or just a "core"? :)

Exactly because we have such a confusing and rapidly changing landscape I think it is unrealistic to have a really common and well defined terminlogy, so once again I think it's better to stick to the terms k8s already uses - and in general focus on our own community and norms -.

fromanirh replied:
in more concrete terms, I'm looking at https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/ and it seems that any possible solution will lead to ambiguity, because the docs and terms leave room for ambiguity. As much as I'd love to clarify, I think there is too much legacy to have a chance to improve in this area.

However the docs also hints at `vCPU`, which could be our best handle here.
My proposal is to rename as `full-pcpus-only` (note "p" for physical cpus) which is again admittedly suboptimal but at least hints at how cpumanager internally represents objects (which is also circling back to initial suggestion from @klueska ) and also hints at the vcpu vs pcpu split.

swatisehgal replied:
Sounds reasonable to me!

ffromani committed:
smtalign: propagate policy options to policies

Consume in the static policy the cpu manager policy options from
the cpumanager instance.
Validate in the none policy if any option is given, and fail if so -
this is almost surely a configuration mistake.

Add new cpumanager.Options type to hold the options and translate from
user arguments to flags.

Co-authored-by: Swati Sehgal <swsehgal@redhat.com>
Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
smtalign: e2e: add tests

Add e2e tests to cover the basic flows for the `full-pcpus-only` option:
negative flow to ensure rejection with proper error message, and
positive flow to verify the actual cpu allocation.

Co-authored-by: Swati Sehgal <swsehgal@redhat.com>
Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/triage accepted
/priority important-soon
because we want this in 1.22 :)

fromanirh commented:
> Does it make sense to split this into two PRs? One for the API change to add `--cpu-manager-policy-options`. And one to add the actual option itself. That way (once the API change is approved), I will have full approval rights for the rest of the changes and we can iterate more quickly, knowing it won't be blocked by the API change.

done here: https://github.com/kubernetes/kubernetes/pull/102055
Let's add the flags to this PR as well and let's discuss with API reviewers which approach (all in one or split PR) they prefer.

fromanirh commented:
/label api-review
/kind api-change

fromanirh commented:
/retest

fromanirh commented:
/retest

ffromani committed:
smtalign: add auto generated code

Files generate after running `make generated_files`.

Co-authored-by: Swati Sehgal <swsehgal@redhat.com>
Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
rebased on top of the changes discussed in https://github.com/kubernetes/kubernetes/pull/102055

fromanirh commented:
/retest

ehashman commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -87,7 +110,14 @@ var _ Policy = &staticPolicy{}
 // container process starts.
-func NewStaticPolicy(topology *topology.CPUTopology, numReservedCPUs int, reservedCPUs cpuset.CPUSet, affinity topologymanager.Store) (Policy, error) {
+func NewStaticPolicy(topology *topology.CPUTopology, numReservedCPUs int, reservedCPUs cpuset.CPUSet, affinity topologymanager.Store, cpuPolicyOptions map[string]string) (Policy, error) {
+	opts, err := NewStaticPolicyOptions(cpuPolicyOptions)
+	if err != nil {
+		return nil, err
+	}
```
This should be surrounded by the feature flag, right? There's no way to get the previous behaviour back otherwise.

fromanirh replied:
We used to have a feature flag check but it was moved earlier in the flow (https://github.com/kubernetes/kubernetes/pull/101432/commits/91e6a8c4fbdd2bbbf607fc53e3fbf1fe705c4ff9). Once we are here, we can't possibly get the options (in the production code) unless we have the right set of feature flags.

klueska replied:
Agree with @fromanirh on the desired flow. All feature flag checks are done in the `containermanager`, with an empty set of options passed in if none are specified or the feature flag is turned off. This is consistent with how similar feature flags are treated for the topologymanager and devicemanager as well.

ehashman commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -220,6 +251,21 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
 		klog.InfoS("Static policy: Allocate", "pod", klog.KObj(pod), "containerName", container.Name)
 		// container belongs in an exclusively allocated pool
 
+		if p.options.FullPhysicalCPUsOnly && ((numCPUs % p.topology.CPUsPerCore()) != 0) {
```
Needs feature flag?

fromanirh replied:
same as above

klueska replied:
ditto

fromanirh commented:
updated and rebased on top of https://github.com/kubernetes/kubernetes/pull/102055

fromanirh commented:
rebased to address conflicts

fromanirh commented:
/test pull-kubernetes-node-e2e-containerd

fromanirh commented:
/retest

fromanirh commented:
```
error during ./hack/e2e-internal/e2e-up.sh: exit status 1
```

fromanirh commented:
/retest
it seems we're still hitting infra issues

fromanirh commented:
since last upload: remove unneeded import rename (overzelous tooling) and added the missing unit test for the `none` policy.

fromanirh commented:
/retest

fromanirh commented:
/retest

fromanirh commented:
/test pull-kubernetes-e2e-gce-alpha-features
it seems a flake (at first glance)

fromanirh commented:
in the gce-alpha-features lane, the cluster fails because the scheduler doesn't go running in the control plane, which is in turn caused by this error
```
profiles[0].pluginConfig[5].args.shape: Required value: at least one point must be specified
```
which is completely unrelated to this PR.

klueska commented:
/approve for the core changes (i.e. not the api or feature changes)

The only open question on this front is whether we want to introduce a _new_ policy to which the new `policy flags` apply, or just have them apply to the existing `static` policy (and any future policies, as appropriate).

This discussion was started because @kad suggested during the KEP review that it might be preferable to not modify the behaviour of the existing `static` policy with these new `policy flags`. It was suggested there is at least one 3rd party tool that relies on the _internal_ behaviour of the `static` policy and only currently keys off of the policy name to see if this feature is turned on or not. With the introduction of these new `policy flags`, this his tool would need to modified to account for them. Otherwise it would risk inferring the wrong behaviour if the `static` policy + some set of `policy flags` was set in a future version of Kubernetes.

It was later clarified, however, that this tool doesn't actually infer any _internal_ behaviour. Instead, it "deploys clusters with specific topology/cpu manager options for AI specific setups". Meaning it decides if some set of `kubelet` flags are appropriate for a given cluster at deployment time.

After having thought about this for a while (and having considered the clarification mentioned above), I’m inclined *not* to introduce a new policy just to accommodate this new options flag.

The reasons being:
1. There is only one such tool we know of that may be affected by this
2. This tool appears to generate the desired set of flags it wants to apply to the `kubelet` (rather than infer the underlying CPU allocation algorithm based on the flags set)
3. If necessary, it should be fairly straightforward to write a conditional statement to check for the presence of the new options flag (without checking specific K8s version numbers).
4. This feature is still in alpha -- if we discover that this is more of an issue than I think it is, we can revisit it before this feature goes beta

Please add comments if you still have any concerns about this.

liggitt commented:
config API changes lgtm (reviewed in https://github.com/kubernetes/kubernetes/pull/102055 and no change in the PR)

fromanirh commented:
/retest
I'm quite sure they are all flakes

fromanirh commented:
/test pull-kubernetes-integration

fromanirh commented:
/test pull-kubernetes-integration

fromanirh commented:
The lane is still failing, but the set of tests failing inside the lane is changing, so I'll report this as flake ASAP.

ehashman reviewed (commented):
/approve
for pkg/features -- all my previous comments were addressed

checked:
- [x] Implement the new policy.
- [x] Ensure proper e2e node tests are in place.

fromanirh commented:
/test pull-kubernetes-integration

klueska commented:
@liggitt all non API approvals are in place now.

liggitt commented:
integration timeout flake is really bad, tracked in https://github.com/kubernetes/kubernetes/issues/103512

liggitt commented:
/approve
for API bits

ehashman commented:
/retest

This still needs a final LGTM.

rphillips commented:
I also reviewed the code and did not see any further additions. It also looks like everything is resolved here.

/lgtm

ehashman commented:
/retest

fromanirh commented:
rebased to fix the conflicts. Thanks for the approvals!

klueska commented:
/lgtm

fromanirh commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

fromanirh commented:
interesting:
![image](https://user-images.githubusercontent.com/25349329/124933078-e36c9580-e003-11eb-9247-f9796a701757.png)

fromanirh commented:
/test pull-kubernetes-integration

fromanirh commented:
/test pull-kubernetes-integration

fromanirh commented:
the integration lane failures seems to be way less likely in the EU (GMT+0/1) morning, aka the NA night. We had some success in the earlier runs today (around 9:00 GMT+1), now is failing badly again.

fromanirh commented:
/retest

fromanirh commented:
/retest

ehashman commented:
Are we certain the integration test failures here are a flake? I haven't seen these sorts of repeated failures on other PRs.

liggitt commented:
> Are we certain the integration test failures here are a flake? I haven't seen these sorts of repeated failures on other PRs.

yes. see https://github.com/kubernetes/kubernetes/issues/103512 and https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&job=integration&test=TestMain

fromanirh commented:
> Are we certain the integration test failures here are a flake? I haven't seen these sorts of repeated failures on other PRs.

Additionally to what Jordan wrote, last three uploads, which are failing pretty constantly, were rebases to address conflicts in `pkg/features`.

derekwaynecarr commented:
@klueska just recording that I agree and feel as I did prior during the KEP review that we did not need an additional flag at this time.  thanks for summarizing above and helping shephard review.

see:
https://github.com/kubernetes/enhancements/pull/2626/files#diff-ba3f042638632df15dfd028bd719d2a645d6f5681b1ccbae355320315e3a96a9R238

ehashman commented:
This is fully approved before the deadline and blocked on a failing test so I don't think it will require an exception, but @fromanirh filed one just in case: https://groups.google.com/g/kubernetes-sig-node/c/19yfkAEAwWU/m/kFaaazlOAwAJ

/milestone v1.22

fromanirh commented:
/retest

fromanirh commented:
rebased to fix the conflict

klueska commented:
/lgtm

fromanirh commented:
thanks @klueska :)

fromanirh commented:
phew, the automated retry made it! It seems https://github.com/kubernetes/kubernetes/issues/103512 is load-related and less likely to happen overnight.
